### PR TITLE
Fix stub for SpooledTemporaryFile

### LINKED ISSUE #2431: SpooledTemporaryFile is a class, not a function
`tempfile.SpooledTemporaryFile` [currently has a stub as a function](https://github.com/python/typeshed/blob/dc27768401b09e22bcd9756ee8639eae06f55bab/stdlib/3/tempfile.pyi#L29), but [it is a class](https://github.com/python/cpython/blob/f9925d86c9d7a44b612576131cd5b84cfee36c4e/Lib/tempfile.py#L630).

See also https://github.com/python/mypy/issues/5586.

### PR DESCRIPTION
**Pull request by ruuda**
Fixes #2431.

Not sure if inheriting from `IO[Any]` is the right way to do it, because in reality the class does not, but it implements the protocol. What is the proper way to indicate that?

Also updates some argument types to include `Optional` explicitly, to match the definition.

### CODE REVIEW

**Comment by srittau** [comment]
At the moment, inheriting from `IO[Any]` is the best way to do that, although in the far, protocol-based IO future this will not be necessary anymore. Unfortunately that also means that you need to copy the abstract methods from `IO` into `SpooledTemporaryFile`.

File: stdlib/3/tempfile.pyi [hunk 1]
```diff
@@ -26,12 +26,20 @@ if sys.version_info >= (3, 5):
...
+    # It does not actually derive from IO[Any], but it does implement the
+    # protocol.
+    class SpooledTemporaryFile(IO[Any]):
+        def __init__(self, max_size: int = ..., mode: str = ...,
+                     buffering: int = ..., encoding: Optional[str] = ...,
+                     newline: Optional[str] = ..., suffix: Optional[str] = ...,
+                     prefix: Optional[str] = ..., dir: Optional[str] = ...
+                     ) -> None: ...
+        def rollover(self) -> None: ...
+        def __enter__(self) -> 'SpooledTemporaryFile': ...
```
Comment (JelleZijlstra) [inline stdlib/3/tempfile.pyi]: Quoting is not necessary in stubs.

Comment (ruuda) [reply stdlib/3/tempfile.pyi]: Fixed.

**Comment by ruuda** [comment]
> Unfortunately that also means that you need to copy the abstract methods from IO into SpooledTemporaryFile.

I copied the abstract methods from `IO` now.

**Comment by srittau** [comment]
Could you have a look at the build failures? It seems that imports of `abstractmethod` and `Iterable` are missing. Also, there' s a linter issue.

**Comment by ruuda** [comment]
Fixed, thanks for the pointers.

File: stdlib/3/tempfile.pyi [hunk 2]
```diff
@@ -26,12 +27,58 @@ if sys.version_info >= (3, 5):
...
+        @abstractmethod
+        def isatty(self) -> bool: ...
+        @abstractmethod
+        def read(self, n: int = ...) -> AnyStr: ...
+        @abstractmethod
+        def readable(self) -> bool: ...
+        @abstractmethod
+        def readline(self, limit: int = ...) -> AnyStr: ...
+        @abstractmethod
+        def readlines(self, hint: int = ...) -> list[AnyStr]: ...
```
Comment (srittau) [inline stdlib/3/tempfile.pyi]: `-> List[AnyStr]` (uppercase L)

File: stdlib/3/tempfile.pyi [hunk 3]
```diff
@@ -26,12 +27,58 @@ if sys.version_info >= (3, 5):
...
-    def SpooledTemporaryFile(
-        max_size: int = ..., mode: str = ..., buffering: int = ...,
-        encoding: str = ..., newline: str = ..., suffix: Optional[AnyStr] = ...,
-        prefix: Optional[AnyStr] = ..., dir: Optional[AnyStr] = ...
-    ) -> IO[Any]:
-        ...
+
+    # It does not actually derive from IO[Any], but it does implement the
+    # protocol.
+    class SpooledTemporaryFile(IO[Any]):
```
[the remark this answers is not in the document]
Comment (ruuda) [reply stdlib/3/tempfile.pyi]: Done.

**Comment by srittau** [comment]
CI is failing, because `List` is not imported.

**Pull request merged**